wolfcrypt/src/port/arm/cryptoCellAes.c

```diff
@@ -0,0 +1,159 @@
+/* cryptoCellAes.c
+ *
+ * Copyright (C) 2019 wolfSSL Inc.
+ *
+ * This file is part of wolfSSL.
+ *
+ * wolfSSL is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * wolfSSL is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA
+ */
+
+#ifdef HAVE_CONFIG_H
+    #include <config.h>
+#endif
+
+#include <wolfssl/wolfcrypt/settings.h>
+#include <wolfssl/wolfcrypt/error-crypt.h>
+
+#ifdef NO_INLINE
+    #include <wolfssl/wolfcrypt/misc.h>
+#else
+    #define WOLFSSL_MISC_INCLUDED
+    #include <wolfcrypt/src/misc.c>
+#endif
+
+#if !defined(NO_AES)
+#include <wolfssl/wolfcrypt/aes.h>
+
+#if defined(WOLFSSL_CRYPTOCELL)
+#include <wolfssl/wolfcrypt/port/arm/cryptoCell.h>
+
+int wc_AesSetIV(Aes* aes, const byte* iv)
+{
+    if (aes == NULL)
+        return BAD_FUNC_ARG;
+
+    if (iv)
+        XMEMCPY(aes->reg, iv, AES_BLOCK_SIZE);
+    else
+        XMEMSET(aes->reg,  0, AES_BLOCK_SIZE);
+
+    return 0;
+}
+
+WOLFSSL_API int wc_AesInit(Aes* aes, void* heap, int devId)
+{
+    if (aes == NULL)
+        return BAD_FUNC_ARG;
+
+    aes->heap = heap;
+    (void)devId;
+
+    memset(&aes->ctx, 0, sizeof(aes->ctx));
+
+    cc310_enable();
+    cc310_Init();
+
+    return 0;
+}
+
+WOLFSSL_API void wc_AesFree(Aes* aes)
+{
+    (void)aes;
+}
+
+int wc_AesSetKey(Aes* aes, const byte* userKey, word32 keylen, const byte* iv,
+                int dir)
+{
+    SaSiError_t result = SASI_OK;
+    SaSiAesIv_t iv_aes;
+
+    if (aes == NULL ||
+       (keylen != AES_128_KEY_SIZE &&
+        keylen != AES_192_KEY_SIZE &&
+        keylen != AES_256_KEY_SIZE)) {
+        return BAD_FUNC_ARG;
+    }
+
+    if (dir != AES_ENCRYPTION &&
+        dir != AES_DECRYPTION) {
+        return BAD_FUNC_ARG;
+    }
+
+    if (dir == AES_ENCRYPTION) {
+        aes->ctx.mode = SASI_AES_ENCRYPT;
+        SaSi_AesInit(&aes->ctx.user_ctx,
+                     SASI_AES_ENCRYPT,
+                     SASI_AES_MODE_CBC,
```

Comment:
We don't know the AES mode here... Shouldn't this be set later? Is this why CBC is only supported? Appears hardware supports ECB, CTR, CMAC, XTS and OFB.